### ISSUE TITLE
Dictionaries return copies of stored TypedArrays

### ISSUE THREAD

**Issue by TokisanGames** [godotengine/godot-cpp#1508]
### Godot version

4.2.2-stable

### godot-cpp version

godot-4.2.2-stable

### System information

Windows 11/64, RTX 3070, Vulkan

### Issue description

The [dictionary documentation](https://docs.godotengine.org/en/stable/classes/class_dictionary.html) and [array documentation](https://docs.godotengine.org/en/stable/classes/class_array.html) both state that that `Dictionaries/Arrays are always passed by reference.`

However I am receiving copies in gdextension when storing arrays indexed via Vector2i.

### TypedArray: copy

```C++
Dictionary test_dict;
TypedArray<Transform3D> ta;
test_dict[Vector2i(0, 0)] = ta;
ta.push_back(Transform3D(1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1));
UtilityFunctions::print("Array A ptr: ", uint64_t(&ta), " size: ", ta.size());

TypedArray<Transform3D> tb = test_dict[Vector2i(0, 0)];
tb.push_back(Transform3D(0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1));
	
UtilityFunctions::print("Array B ptr: ", uint64_t(&tb), " size: ", tb.size());
UtilityFunctions::print("Array A ptr: ", uint64_t(&ta), " size: ", ta.size());
```

```
Array A ptr: 81925230416 size: 1
Array B ptr: 81925230424 size: 2    // note different memory address
Array A ptr: 81925230416 size: 1   // should be size 2
```

### Array: reference

```C++
Dictionary test_dict2;
Array aa;
test_dict2[Vector2i(0, 0)] = aa;
aa.push_back(Transform3D(1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1));
UtilityFunctions::print("Array A ptr: ", uint64_t(&aa), " size: ", aa.size());

Array ab = test_dict2[Vector2i(0, 0)];
ab.push_back(Transform3D(0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1));

UtilityFunctions::print("Array B ptr: ", uint64_t(&ab), " size: ", ab.size());
UtilityFunctions::print("Array A ptr: ", uint64_t(&aa), " size: ", aa.size());
```

```
Array A ptr: 87499460384 size: 1
Array B ptr: 87499460400 size: 2
Array A ptr: 87499460384 size: 2
```

### Image: reference

```C++
Dictionary test_image;
Ref<Image> img = Image::create(1, 1, true, Image::FORMAT_RGB8);
img->set_pixel(0, 0, Color(1, 1, 1));
UtilityFunctions::print("Image A ptr: ", uint64_t(img.ptr()), " (0,0): ", img->get_pixel(0, 0));
test_image[Vector2i(0, 0)] = img;

Ref<Image> b = test_image[Vector2i(0, 0)];
b->set_pixel(0, 0, Color(.5, .5, .5));
UtilityFunctions::print("Image B ptr: ", uint64_t(b.ptr()), " (0,0): ", b->get_pixel(0, 0));
UtilityFunctions::print("Image A ptr: ", uint64_t(img.ptr()), " (0,0): ", img->get_pixel(0, 0));
```

```
Image A ptr: 3032189040112 (0,0): (1, 1, 1, 1)
Image B ptr: 3032189040112 (0,0): (0.498, 0.498, 0.498, 1)  // same memory pointer
Image A ptr: 3032189040112 (0,0): (0.498, 0.498, 0.498, 1)  // correct value
```

------
I replaced the Godot dictionary with std::map and it works properly with Godot TypedArrays, so I am using that for now. Using a local map while building, then converting to a Godot dictionary for storage within a Godot resource. It's a bit tacky.

**Comment by AThousandShips** [comment]
Does this happen when not using typed arrays? Given your example uses them only

**Comment by TokisanGames** [comment]
Arrays also copied, Images by reference. I have simplified the code above and made it testable.

**Comment by AThousandShips** [comment]
Typed arrays are already tracked here:
* https://github.com/godotengine/godot-cpp/issues/1149

So let's focus on dictionary

**Comment by TokisanGames** [comment]
Please review my changes above. It affects Arrays and TypedArrays. Not Image.

This is likely related to that other issue. It mentions an Array constructor that makes a copy. Perhaps Dictionary is triggering that constructor. The proposed PR only touches TypedArray so I'm guessing it won't fix this issue for Arrays in Dictionaries.

**Comment by ajreckof** [comment]
```
Dictionary test_dict2;
Array aa;
test_dict2[Vector2i(0, 0)] = aa;
aa.push_back(Transform3D(1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1));
UtilityFunctions::print("Array A ptr: ", uint64_t(&aa), " size: ", aa.size());

TypedArray<Transform3D> ab = test_dict2[Vector2i(0, 0)];
ab.push_back(Transform3D(0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1));

UtilityFunctions::print("Array B ptr: ", uint64_t(&ab), " size: ", ab.size());
UtilityFunctions::print("Array A ptr: ", uint64_t(&aa), " size: ", aa.size());
```
would always trigger the constructor (and therefore a copy) since you are implicitly creating a typed array from an array can you try with arrays only something like :

```
Dictionary test_dict2;
Array aa;
test_dict2[Vector2i(0, 0)] = aa;
aa.push_back(Transform3D(1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1));
UtilityFunctions::print("Array A ptr: ", uint64_t(&aa), " size: ", aa.size());

Array ab = test_dict2[Vector2i(0, 0)];
ab.push_back(Transform3D(0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1));

UtilityFunctions::print("Array B ptr: ", uint64_t(&ab), " size: ", ab.size());
UtilityFunctions::print("Array A ptr: ", uint64_t(&aa), " size: ", aa.size());
```

**Comment by TokisanGames** [comment]
@ajreckof Perhaps you saw an outdated version of this ticket. Please refresh and see that I tested TypedArray (copy), Array (copy) and Image (reference).

**Comment by ajreckof** [comment]
I tested with gdscript and can't reproduce so this is gdextension only

**Comment by AThousandShips** [comment]
> Ok, given that it's only related to TypedArray, I think this probably is just a duplicate. AThousandShips do you think your PR will fix this ticket?

Feel free to try it out, I'm not sure, but it does sound like a duplicate as it doesn't affect Array

**Comment by TokisanGames** [comment]
The PR does fix the issue with TypedArray. I can use std::map for now until 4.3 is stable. Thank you both for the help. This was a frustrating journey. The function I was writing now processes 230k instances in .34 seconds and runs ~1700x faster with the PR and not having to writing back to the dictionary.

Duplicate #1149
Fixed by #1483
